File: bench/testdata/samples/A08/PSB-A08-DESER-001/javascript/secure.js
// PSB-A08-DESER-001: User data deserialization
// CWE: CWE-502
// Expected: (none - secure)
const express = require("express");
const app = express();
app.use(express.json());

const ALLOWED_KEYS = new Set(["theme", "language", "timezone", "notifications", "font_size"]);

app.post("/api/preferences/import", (req, res) => {
  const { preferences, user_id } = req.body;
  if (!preferences) {
    return res.status(400).json({ error: "preferences data is required" });
  }

  let parsed;
  try {
    parsed = typeof preferences === "string" ? JSON.parse(preferences) : preferences;
  } catch (err) {
    return res.status(400).json({ error: "invalid JSON preferences" });
  }

  if (typeof parsed !== "object" || Array.isArray(parsed)) {
    return res.status(400).json({ error: "preferences must be an object" });
  }

  const safePrefs = {};
  for (const [key, value] of Object.entries(parsed)) {
    if (ALLOWED_KEYS.has(key)) {
      safePrefs[key] = value;
    }
  }

  // db.query("UPDATE users SET prefs = $1 WHERE id = $2", [JSON.stringify(safePrefs), user_id]);
  res.json({ status: "imported", preferences: safePrefs });
});

module.exports = app;

File: bench/testdata/samples/A08/PSB-A08-DESER-001/python/secure.py
# PSB-A08-DESER-001: User data deserialization
# CWE: CWE-502
# Expected: (none - secure)
from flask import Flask, request, jsonify
import json

app = Flask(__name__)

ALLOWED_KEYS = {"theme", "language", "timezone", "notifications", "font_size"}
VALID_THEMES = {"light", "dark", "system"}
VALID_LANGUAGES = {"en", "es", "fr", "de", "ja", "zh"}

@app.route("/api/preferences/import", methods=["POST"])
def import_preferences():
    data = request.get_json()
    raw = data.get("preferences")
    if not raw:
        return jsonify({"error": "preferences data is required"}), 400

    try:
        if isinstance(raw, str):
            preferences = json.loads(raw)
        else:
            preferences = raw
    except (json.JSONDecodeError, TypeError):
        return jsonify({"error": "invalid JSON preferences"}), 400

    safe_prefs = {k: v for k, v in preferences.items() if k in ALLOWED_KEYS}

    user_id = data.get("user_id")
    # db.execute("UPDATE users SET prefs = ? WHERE id = ?", (json.dumps(safe_prefs), user_id))

    return jsonify({"status": "imported", "preferences": safe_prefs})

if __name__ == "__main__":
    app.run()
